### DISCUSSION TITLE
Skeletal Animation - Constructing Rigid Skeleton from Armature and Animating

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#89] [Q&A]
I have the initial "Create Rig" sample ported over to Jolt Physics JS, but have been working on the "Powered" versions using animations I've baked into some simple GLTF. Since we haven't been hosting static files in the examples, I will probably strip the data out and convert it to JSON

![2024-01-05-SkeletalRig](https://github.com/jrouwe/JoltPhysics.js/assets/756488/d8c2b398-7540-46d4-b7d5-30667dc6c6d0)

1. I'm currently mentally stuck a little in auto constructing the constraints off of a standard Mat44 armature tree, but may figure that out.
I'm assuming in theory you could just grab the chain of Mat44's that correspond to the bones, and take each parent-child pair to get the local-space offsets of each one and use those to build the TwoBody binding.

I'd personally been planning to make a trivial snow man animate/bounce himself into a pile of blocks. I'd need to tag extra data in my armature to determine what 6DOF or other joints are needed, now that I see we can arbitrarily pick any TwoBody constraint per RagdollPart->mToParent.

2. With regard to animating, I'm trying to understand how to, for example, make the snowman bounce forward and hit a stack of blocks. I see there are two methods to do this, and am not 100% clear on them. They seem to be similar to Dynamic vs Kinematic. I think to start off with, I'm planning baking the entire motion into a single animation including full translation, since otherwise a chain of animations (like hopping 3 times towards an object) I'm assuming would require you force-change the root location between each animation loop, probably by stealing it off the last-frame's calculated root Mat44 position.

Assuming it is the Kinematic method of animation, I believe if I use the DriveToPoseUsingKinematics powered version then it may essentially make the snowman an unstoppable force in the world and it would follow it's animation data perfectly and go through the blocks with no resistance.

Assuming it is the Dynamic method of animation, I assume if I used DriveToPoseUsingMotors, then it seems like each joint of the skeleton will just apply X force of the motors to try its best to force every bone to the right Position/Rotation in world space. So in this case, I think it would kind of bounce off the blocks itself a little, and Bodies too heavy will actually stop it from hitting the target pose.

3. I'm still not 100% sure how or why the model won't kind of fall over and have the animation be unable to upright itself to continue walking. Would I need to tie all roots to a Kinematic so the root orientation won't allow a walking character to tip?

Once any animation finishes. I'm assuming that I'd need to flip to an idle pose and continually drive that animation pose of a character standing in place, or it will just collapse pure-ragdoll. 

![2024-01-05-Snowman-Rig-Blender](https://github.com/jrouwe/JoltPhysics.js/assets/756488/c74fa07c-8804-4ed3-bdb2-0e4cee49850b)

Some of these may resolve themselves when I work on loading and auto-rigging a model.

**Comment by jrouwe** [comment]
> I have the initial "Create Rig" sample ported over to Jolt Physics JS.

Nice!

> I'm currently mentally stuck a little in auto constructing the constraints off of a standard Mat44 armature tree, but may figure that out.
I'm assuming in theory you could just grab the chain of Mat44's that correspond to the bones, and take each parent-child pair to get the local-space offsets of each one and use those to build the TwoBody binding.

I'm not sure if I understand the problem exactly, but you can create the constraints either in local or in world space (see SwingTwistConstraintSettings.mSpace), so in principle you don't need to do the conversion when creating a rig as Jolt will do it for you.

> With regard to animating, I'm trying to understand how to, for example, make the snowman bounce forward and hit a stack of blocks. I see there are two methods to do this, and am not 100% clear on them.

This is not the easiest example to get working. What I would probably do is to create the rigid bodies as Dynamic with GravityFactor = 0 and call DriveToPoseUsingKinematics to make the snowman perfectly follow the animation. Then on the first contact with a dynamic body I would set GravityFactor = 1 and switch to driving the snowman pose in local space using DriveToPoseUsingMotors (which would mean that from that point onwards the animation of the root joint doesn't do anything anymore and its purely physics).

**Comment by LeXXik** [comment]
Sweet! That would be great to have!

Personally, I'm fine with a repo having a static assets folder with a few assets to make the examples look better. We have a similar one in PlayCanvas for use in the demos:
https://github.com/playcanvas/engine/tree/main/examples/assets
e.g. a `models/chess-board.glb`:
https://playcanvas.github.io/#/graphics/multi-view
It makes it easier to show-off some engine features better (just add a license text file with links somewhere). It does require having some sort of an importer, though. PlayCanvas has a GLTF parser, but it depends on PC, so probably not going to be suitable here, given the examples are using Three:
`src/framework/parsers/glb-parser.js`

About the rig - it reminds me of one I did a while back. It was simpler, as I did not use motors. I was using the kinematic to match animation perfectly, then turning to dynamic for ragdoll, and some skin blending.

https://github.com/jrouwe/JoltPhysics.js/assets/5677782/4b1982d8-4f73-4b91-bdc4-86f801300a3a

I think the motors should help keeping it up straight, when it is all dynamic, even with gravity. However, if it falls, then it won't be able to stand up without some help. In games, usually there is a "push up" motion, that adds an up impulse and helps it to get back to its feet (or use a kinematic stand up animation, like in my clip).

**Reply by PhoenixIllusion** [reply]
Three has a GLTF loader in the examples folder (Three's addon library, features not included by default - like model loading examples/jsm/loaders). I just put it next to the OrbitalControl (also an addon, under /controls/) and wrapped it the same way as the OrbitalControl. It has worked as expected, so far. I just chose the file out of GitHub under the Three tag r144, like the rest of the included Three files.

It should be parsing the animations fine, so I will try the various above recommendations. Likely I will ignore the actual model Shapes on the rig at first and just traverse the armature itself into a Skeleton for my first pass at the demo, then support toggling to a second Skeleton of the model Shapes once I have that working

**Comment by PhoenixIllusion** [comment]
I think it may require a bit more work for me to figure out how Three and GLTF are mapping the bones exactly (or I'm setting up the collision bodies wrong) ... One of the more glaring issues in my model appears to be that my Body items are getting re-rotated and possibly repositioned by initializing the body to a position/rotation, and then applying the skeletal transform, which then possibly causes it to wipe the old rotations. I think if I use rotated/translated shape settings I can properly bake the first position/rotation transform into the shape itself so that the BodyInterface just drives the bone's animation  transform instead of also wiping some parts of the Bind Pose mat44 data.

From the left:
1st is a debug of just pure bone data with rotations
2nd is the ragdoll with Kinematic (actual Jolt controlled model). Trying to do Dynamic or Drive with Motors just pushes against the constraints wrong since I think the forces are essentially trying to break the skeletons arms and legs
3rd is the collision model used to build the ragdoll, just boxes placed using Blender over the bind-pose.
4th is the real model I made and then used Adobe's Mixamo site to animate for example cases on my Jolt/BabylonJS 3rd person demos.

![2024-01-06-Character-Rig](https://github.com/jrouwe/JoltPhysics.js/assets/756488/edd6d8b9-f058-4b2d-ba0d-55bd8a25a087)

Mixamo does have a more limited license than something like CCO, so I may look into alternative GLTF for the final "powered rig" demo. The model is original work I can release claim to or whatever, but the animation data is the only tricky thing. Their terms are that you can freely use the model with animations data in personal/commercial games/products/art per Adobe's terms of the animation data, and I believe attributing that it uses Adobe Mixamo and directing to their license would cover that, but the more nebulous topic is situation where it is "not inside a game or application" like being in the source of a demo. I know direct resale of the animation data or direct export an distribution of the raw animation data outside of a model may not be covered.

Likely there is some CCO or other open model, or a go with a 100% custom made GLTF with animation that I could use and release if we were to include an asset file.

**Comment by jrouwe** [comment]
>  I think if I use rotated/translated shape settings I can properly bake the first position/rotation transform into the shape itself so that the BodyInterface just drives the bone's animation transform instead of also wiping some parts of the Bind Pose mat44 data.

This is a common approach, so yes I would go for that.

> Likely there is some CCO or other open model, or a go with a 100% custom made GLTF with animation that I could use and release if we were to include an asset file.

You could also take the [rig](https://github.com/jrouwe/JoltPhysics/blob/master/Assets/Human.tof) and [animations](https://github.com/jrouwe/JoltPhysics/tree/master/Assets/Human) that the Jolt sample app uses and convert them to a JS compatible format, that way the license will be covered. They can be read by the C++ code and it should be quite easy to write a bit of code that spits them out in another text format (or alternatively you parse them with your own script code).

**Reply by PhoenixIllusion** [reply]
That is probably enough of a task to get a PR with the required SDK code in first and direct ports of samples, and then a later add an example with how to apply it to arbitrary models. I would probably keep this GitHub discussions alive until the custom model part is figured out, but the 1st PR is a good checkpoint.

**Reply by PhoenixIllusion** [reply]
The trivial route would be wrapping the std:string in a string_stream exposed via IDL, so it would allow as-is loading of files without any custom code, just construct a std::string of the fetch() => text() AJAX and then pass that to a string-stream constructor and load it as a stream-input of the ObjectStreamIn.

For now I've made a simple parser of the file format and will see if that is easy to use as well
![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/926ab448-3bb2-4ce8-84e7-49464aa94f44)

**Reply by PhoenixIllusion** [reply]
Populating the GroupFilterTable via JS parsed ObjectStream data without access to private mTable  is a bit tricky to setup, and I need to confirm extra physics on the "Load Rig", but the Human.tof file is loading (I think, I'm not sure if the face should look like this or not?)

![2024-01-08-Load-Rig](https://github.com/jrouwe/JoltPhysics.js/assets/756488/6725965a-1a12-4ba5-b7ab-f25040ce83cf)

**Comment by PhoenixIllusion** [comment]
I at least have the TOF Kinematic Rig example loading. Kind of looks caffeinated when you force the rig to a single 1-frame pose every frame. There is also some slight issue with animations snapping the model back to initial location and obliterating any objects in their way, but that is kind of expected.

![2024-01-08-Kinematic_Rig](https://github.com/jrouwe/JoltPhysics.js/assets/756488/fb0085d5-0f8c-44b7-9c9d-6dcbb41e9619)

**Reply by PhoenixIllusion** [reply]
Oh, that was just that I'd loaded the ragdoll as Dynamic and then powered by Kinematic, so the forces were overpowering the model.

**Reply by PhoenixIllusion** [reply]
Not sure how to get Powered Rig to work correct. I can see minor motion on the body, but it is just a slight wiggle.
![2024-01-08-Powered_Rig](https://github.com/jrouwe/JoltPhysics.js/assets/756488/08582db1-738a-46b2-8b4e-95533f850d5f)

**Comment by jrouwe** [comment]
> The trivial route would be wrapping the std:string in a string_stream exposed via IDL, so it would allow as-is loading of files without any custom code, just construct a std::string of the fetch() => text() AJAX and then pass that to a string-stream constructor and load it as a stream-input of the ObjectStreamIn.

I'm fine with this solution.

> I think, I'm not sure if the face should look like this or not?

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/0b1875dc-6c10-44dd-bace-3b6033c357c5)

The neck looks a bit odd here as well :)

> Populating the GroupFilterTable via JS parsed ObjectStream data without access to private mTable is a bit tricky to setup

I would prefer not to expose mTable as it is too much tied to the implementation. In this case you can also get away by just calling RagdollSettings.DisableParentChildCollisions (inJointMatrices = nullptr should work fine for this ragdoll, but you could pass in the bind pose for a slightly better result).

> Not sure how to get Powered Rig to work correct. I can see minor motion on the body, but it is just a slight wiggle.

This is a bit hard to debug without any code. If you send me a PR I can integrate what you've done so far and I can take a look at what could be wrong. I'm assuming you're calling DriveToPoseUsingMotors in this example. Possibly the frequency of the motors is too low (try 20 for MotorSettings.mSpringSettings.mFrequency) or the damping too high (try something close to 0) or the torque limits could be too low (MotorSettings.mMinTorqueLimit, mMaxTorqueLimit), note that min_limit should be -max_limit and FLT_MAX is a good option to start with for max_limit.

**Reply by PhoenixIllusion** [reply]
Spring settings looks like it fixed it, though it can't stand on it's own.

The existing spring/constraint data was pulled directly out of the TOF on those, but the TOF doesn't set variables for spring settings. Changing it like you said fixed the issue, though I have that code now in the Rig Loader JS itself vs just the powered_rig.

I can prep a Squashed PR of this. Do you think it is an issue if I put the 4x Rig html in their own folder vs just a giant root one?
I have ports of these 4
* create_rig
* load_rig
* kinematic_rig
* powered_rig

I started doing big_world but that seemed more draw-distance related. I looked into rig_pile but don't have the map-loading for the BOF format yet. I suppose I could do that one with the above string_stream logic since my JS TOF-parser doesn't understand it.

![2024-01-08-Powered_Rig2](https://github.com/jrouwe/JoltPhysics.js/assets/756488/5ab1585b-64d4-43d2-afee-9f8804bc0151)

**Reply by PhoenixIllusion** [reply]
string_stream of the BOF kind of works, but way too oversized (5000+ items with giant height-maps) for our tutorial examples without allowing memory growth and overwhelming our automatic THREE setup. The height-field auto-generation code also will trigger the super-slow default addToScene that I had to code around for in the other HeightField demo where it does a poorly optimized giant mesh instead of just the GPU accelerated displacement map that I ended up using.

**Reply by PhoenixIllusion** [reply]
![2024-01-08-BOF-Load-Terrain1](https://github.com/jrouwe/JoltPhysics.js/assets/756488/fd69c0b5-5a30-4e33-970c-afb91aac47f6)

Some of the shapes in it besides heightmap also cause GetTriangle to throw memory-out-of-bound exceptions, but I think I will skip on these model things unless I can find a way to convert the data structure for rendering into something nicer for THREE.

**Reply by PhoenixIllusion** [reply]
![2024-01-08-BOF-Load-Terrain1-first-1000](https://github.com/jrouwe/JoltPhysics.js/assets/756488/31789c3e-76e4-4b71-ab18-0e21bab93577)
pretty though

**Reply by jrouwe** [reply]
I've merged the PR and released a new version with the changes.

> Do you think it is an issue if I put the 4x Rig html in their own folder vs just a giant root one?

I think it's fine. It's probably time to organize the others as well. I didn't anticipate that I would get this many demos.

> I started doing big_world but that seemed more draw-distance related.

It also shows a simulation issue, but replicating the demo would require compiling the library with double support (define JPH_DOUBLE_PRECISION) and changing a whole bunch of functions and attributes in Jolt.idl from Vec3 to RVec3 (and go through all demos to do the same). My original thought was that it would be unlikely for anyone to ever make a game that's bigger than 4 x 4 km in a browser, so I just made float the default.

> pretty though

Yup! But I'm not sure how much effort this is worth. Loading a BOF file is probably something nobody will ever do and the demo doesn't really show anything new either.

**Reply by PhoenixIllusion** [reply]
I got the BOF file to load by throwing together a quick JS binary parser for it using DataView. I can put it in a demo as at least a playground people could load up and throw things in (once i'm available again, going to India). I had to assume when to load a Instance ID vs nested Class, so for now just hard-coded it to flag off what I thought were abstract-style classes (GroupFilter, Shape, PhysicsMaterial). That seemed to be enough to get it to fully parse the terrain1.bof and terrain2.bof

I decided since it is all static content, its probably easiest to just load all the data to Jolt, and then render the map using pure graphical code to more easily share geometry and do tricks like drops all compound elements to their sub-shapes so it can properly frustrum cull.

I'm not sure how to properly sync the HeightMap to the rest of the level though with what time I have available at this instant.

This load all  elements in the level and renders the entire level (just using Three so far) at 30fps+ on my machine, so not horrible.
https://github.com/PhoenixIllusion/JoltPhysics.js/tree/feature/bof-loading

Probably oct-tree frustrum culler and some Potential Visibility algorithms would help cull better, overkill for this demo.

![2024-01-09-BOF-Load-Terrain1](https://github.com/jrouwe/JoltPhysics.js/assets/756488/9b975b8a-64b3-4a4b-852a-00a9444e9ff0)

**Reply by PhoenixIllusion** [reply]
Flip along x/y diagonal for the height-map causes THREE to render it as expected and line up under all the decorations
![2024-01-10-BOF-Load-Terrain3](https://github.com/jrouwe/JoltPhysics.js/assets/756488/fa8cbe6c-6799-4b7c-b9df-06f0cc4e2fed)
![2024-01-10-BOF-Load-Terrain2](https://github.com/jrouwe/JoltPhysics.js/assets/756488/30168087-4a45-4888-a4a3-dc7fd8b3ce85)

**Reply by jrouwe** [reply]
Nice! The height field looks a bit smoother here though. I'm guessing you quantize to 8 bits?

![image](https://github.com/jrouwe/JoltPhysics.js/assets/1621693/a61b34a8-0329-4405-abc2-25d4bbedde5f)

**Reply by jrouwe** [reply]
Quickly scanning through your bof-loading branch: Instead of manually rendering convex hulls you can also 'unpack' the compound shapes by calling ```body.GetTransformedShape().CollectTransformedShapes(...)```. That way you would only need to check if the shape is a height field. For other shapes you can probably just use the default renderer.

**Reply by PhoenixIllusion** [reply]
I found a way afterword to use the height map data and transform the plane to use the float heights, which looks much smoother. It looks much more like your reference.

For the mesh shapes, I prefer to use the native THREE one so it uses index variables.

In general I'd been trying to avoid the GetTriangles related calls so I could process the sub shapes more and find cases of possible MeshInstancing or splitting out the meshes so they could be structured under a hierarchy and help the culling.

**Reply by LeXXik** [reply]
What's the village name? I should visit it in-game :) 
I want to try to load in PlayCanvas as well. Will give it a go this weekend.

**Comment by jrouwe** [comment]
Ok.

Note that I just merged your changes to Jolt.h/idl.

**Comment by LeXXik** [comment]
I think this is a great progress! Thank you for tackling this!

I should probably mention that the rigging example at the current state (even though a great technical feat) is a bit convoluted for an example. I'm just not sure how useful it will be for the web dev. I hope nobody will start looking into how to export their animations to .tof format, so that they can add physics to it.

I guess I'm trying to say that we should be careful with increasing the library size with features that people won't be able to use. Can RagdollLoader be used with more common animation assets, like FBX/GLB?

**Reply by PhoenixIllusion** [reply]
That avoidance of increasing the library is another reason why I have coded the RagDoll Loader and TOF/BOF code as pure stand-alone JS, interior to the Example folder, instead of using the emscripten versions or anything linked to the NPM related JS files. By not using the WebIDL code for the parsing, they do not increase the WASM library size with parsers or other unused functionality, only we added what is required in the core lib for purely generic animation of rigs, and also give a starting point for what is required for porting future animations if you want to code FBX/GLB/GLTF models their own physics animations.

My initial stab was to animate my GLB, but I instead decided to get a guaranteed working version of the TOF file first. I may work on the rigging of my GLB again sometime soon, but have some other proc-gen code and go-kart setups I'm doing with my Babylon/Jolt as well.

Per guidance on Bullet/Ammo, if the WebIDL/Wasm size does get to be an issue on project that require a smaller payload when linking in, the easiest route is commenting out chunks of the WebIDL and rebuilding a custom build with just the required methods for that project. *edited - spelling and phrasing, posted while flying in plane during takeoff, losing cell data
